src/main/java/dan200/computercraft/shared/peripheral/generic/data/DetailProviders.java

```
@@ -0,0 +1,79 @@
+
+    public static synchronized <T> void register( IDetailProvider<T> provider )
+    {
+        Objects.requireNonNull( provider, "provider cannot be null" );
+
+        // A detail provider can implement multiple interfaces. Check the valid ones, and log an error if the given
+        // provider matched none of them.
+        boolean matched = false;
+
+        if ( provider instanceof IBlockDetailProvider )
+        {
+            registerProvider( BlockState.class, (IBlockDetailProvider) provider );
+            matched = true;
+        }
+
+        if ( provider instanceof IFluidDetailProvider )
+        {
+            registerProvider( FluidStack.class, (IFluidDetailProvider) provider );
+            matched = true;
+        }
+
+        if ( provider instanceof IItemDetailProvider )
+        {
+            registerProvider( ItemStack.class, (IItemDetailProvider) provider );
+            matched = true;
+        }
+
+        if ( !matched )
+        {
+            ComputerCraft.log.error( "Detail provider {} does not implement any valid interfaces.", provider.getClass().getName() );
+        }
+    }
+
+    private static <T> void registerProvider( Class<T> type, IDetailProvider<T> provider )
+    {
+        allProviders.computeIfAbsent( type, k -> new LinkedHashSet<>() ).add( provider );
+    }
+
+    @SuppressWarnings( "unchecked" )
+    public static <T> void fillData( Class<T> type, Map<? super String, Object> data, T value )
+    {
+        Collection<IDetailProvider<T>> providers = (Collection<IDetailProvider<T>>) (Object) allProviders.get( type );
+        if ( providers == null ) return;
+
+        for ( IDetailProvider<T> provider : providers )
+        {
+            try
+            {
+                provider.provideDetails( data, value );
+            }
+            catch ( Exception e )
+            {
+                ComputerCraft.log.error( "Error while providing details for {}", value, e );
```

I know we do this for peripherals, but I think we should probably just bubble the exception. Better to signal an error to the computer than silently drop data.